### PR TITLE
Anagram finder

### LINKED ISSUE #40: Anagram Finder
The Anagram Finder challenge involves creating a function to determine whether two given strings are anagrams. An anagram is a word or phrase formed by rearranging the letters of another, typically using all the original letters exactly once. For instance, "listen" and "silent" are anagrams because their letters can be rearranged to form one another. The challenge also accounts for case insensitivity and ignores spaces, allowing phrases like "Dormitory" and "Dirty Room" to be considered valid anagrams. The goal is to implement a solution that efficiently checks for anagrams while adhering to these criteria.

### PR DESCRIPTION
**Pull request by Frank2446-dotcom**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/anagram_finder.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
+"""
+Anagram Finder
+This script provides a function to determine if two strings are anagrams.
+"""
+
+
+def is_anagram(string1: str, string2: str) -> bool:
+    """
```
Comment (Derekkarungani) [inline solutions/anagram_finder.py:13]: Add at least 3 doctests to ensure the function has sufficient examples for testing and documentation purposes.

File: solutions/tests/test_anagram_finder.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+class TestAnagramFinder(unittest.TestCase):
+    """Test cases for the is_anagram function."""
+
+    def test_anagrams(self):
+        """Test cases for valid anagrams."""
+        self.assertTrue(is_anagram("listen", "silent"))
+        self.assertTrue(is_anagram("evil", "vile"))
+        self.assertTrue(is_anagram("dusty", "study"))
+        self.assertTrue(is_anagram("night", "thing"))
+        self.assertTrue(is_anagram("brag", "grab"))
```
Comment (Derekkarungani) [inline solutions/tests/test_anagram_finder.py]: Well-structured and comprehensive unit test code. Great job.

File: solutions/anagram_finder.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+It checks if the strings are anagrams, handling edge cases like empty strings, special characters, and case-insensitivity.
+
+Created on 03 01 2025
+@author: Frankline Ambetsa
+"""
+
+import doctest
+
+
+def is_anagram(string1: str, string2: str) -> bool:
```
Comment (Kareiman) [inline solutions/anagram_finder.py]: The function name should match the file name as per the checklist requirements

File: solutions/anagram_finder.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        raise AssertionError("string2 cannot be empty")
+
+    # Remove spaces and convert to lowercase
+    string1 = string1.replace(" ", "").lower()
+    string2 = string2.replace(" ", "").lower()
+
+    # Check if sorted strings are equal (anagram check)
+    return sorted(string1) == sorted(string2)
+
+
```
Comment (Kareiman) [inline solutions/anagram_finder.py:61]: As per the code checklist, you should not call the function inside the code

File: solutions/tests/test_anagram_finder.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
+# Import statements must be at the top of the file
+import unittest
+
+from solutions.anagram_finder import is_anagram
```
Comment (Kareiman) [inline solutions/tests/test_anagram_finder.py]: You may change the function name her after you rename it to match the file name

**Review by Kareiman — APPROVED** [review]
Great work @Frank2446-dotcom 
Detailed work and well-structured code. Well-done!

**Review by Derekkarungani — APPROVED** [review]

File: solutions/anagram_finder.py (hunk 4 above)
Comment (Frank2446-dotcom) [reply solutions/anagram_finder.py:61]: Done

File: solutions/tests/test_anagram_finder.py (hunk 5 above)
Comment (Frank2446-dotcom) [reply solutions/tests/test_anagram_finder.py]: Done
